iroh/src/endpoint.rs

```diff
@@ -2183,4 +2262,110 @@ mod tests {
             .unwrap()
             .unwrap();
     }
+
+    #[tokio::test]
+    #[traced_test]
+    async fn test_connect_with_opts_0rtt() -> TestResult {
+        async fn spawn_server() -> Result<Endpoint> {
+            let server = Endpoint::builder()
+                .alpns(vec![TEST_ALPN.to_vec()])
+                .relay_mode(RelayMode::Disabled)
+                .bind()
+                .await?;
+
+            // Gets aborted via the endpoint closing causing an `Err`
+            tokio::spawn({
+                let server = server.clone();
+                async move {
+                    while let Some(incoming) = server.accept().await {
+                        let connecting = incoming.accept()?;
+                        let conn = match connecting.into_0rtt() {
+                            Ok((conn, _)) => {
+                                info!("0rtt accepted");
+                                conn
+                            }
+                            Err(connecting) => {
+                                info!("0rtt denied");
+                                connecting.await?
+                            }
+                        };
+                        let mut recv = conn.accept_uni().await?;
+                        let _ = recv.read_to_end(1_000).await?;
+
+                        let mut send = conn.open_uni().await?;
+                        send.write_all(b"thank you").await?;
+                        send.finish()?;
+
+                        // Gracefully close the connection
+                        conn.closed().await;
+                    }
+                    anyhow::Ok(())
+                }
+            });
+
+            Ok(server)
+        }
+
+        async fn connect_client_expect_0rtt_err(
+            client: &Endpoint,
+            server_addr: NodeAddr,
+        ) -> Result<()> {
+            let conn = client
+                .connect_with_opts(server_addr, TEST_ALPN, ConnectOptions::new())
+                .await?
+                .into_0rtt()
+                .expect_err("0rtt doesn't have keys on first connection yet")
+                .await?;
+
+            let mut send = conn.open_uni().await?;
+            send.write_all(b"hello").await?;
+            send.finish()?;
+            let mut recv = conn.accept_uni().await?;
+            recv.read_to_end(1_000).await?;
+            conn.close(0u32.into(), b"thx");
+            Ok(())
+        }
+
+        async fn connect_client_expect_0rtt_ok(
+            client: &Endpoint,
+            server_addr: NodeAddr,
+        ) -> Result<()> {
+            let (conn, zero_rtt_accepted) = client
+                .connect_with_opts(server_addr, TEST_ALPN, ConnectOptions::new())
+                .await?
+                .into_0rtt()
+                .map_err(|_| "0rtt failed")
```

Comment:
You could merge these two client functions by adding an extra param to indicate whether 0-rtt is expected to work?  It seems they are otherwise identical so might be easier to maintain.

Reply:
I tried merging the functions, but IMO the maintenance effort actually goes up when you realize how weird the function ends up looking. ("What do I do with the `assert!(zero_rtt_accepted.await)`? do I replace that with a dummy value if I don't expect 0-RTT to work? If I replace that with a dummy value, what type does it have? `Either<ZeroRttAccepted, future::Ready<bool>>`? What is the name of that variable, given it's not really checking that 0RTT was accepted?" etc.)
It just gets more confusing. Doing the same mechanical change in two places should e.g. the stream API change is much less confusing.